README.md
# NAME

Struct::Path::PerlStyle - Perl-style syntax frontend for [Struct::Path](https://metacpan.org/pod/Struct::Path).

<a href="https://travis-ci.org/mr-mixas/Struct-Path-PerlStyle.pm"><img src="https://travis-ci.org/mr-mixas/Struct-Path-PerlStyle.pm.svg?branch=master" alt="Travis CI"></a>
<a href='https://coveralls.io/github/mr-mixas/Struct-Path-PerlStyle.pm?branch=master'><img src='https://coveralls.io/repos/github/mr-mixas/Struct-Path-PerlStyle.pm/badge.svg?branch=master' alt='Coverage Status'/></a>
<a href="https://badge.fury.io/pl/Struct-Path-PerlStyle"><img src="https://badge.fury.io/pl/Struct-Path-PerlStyle.svg" alt="CPAN version"></a>

# VERSION

Version 0.80

# SYNOPSIS

    use Struct::Path qw(spath);
    use Struct::Path::PerlStyle qw(path2str str2path);

    my $nested = {
        a => {
            b => ["B0", "B1", "B2"],
            c => ["C0", "C1"],
            d => {},
        },
    };

    my @found = path($nested, str2path('{a}{}[0,2]'), deref => 1, paths => 1);

    while (@found) {
        my $path = shift @found;
        my $data = shift @found;

        print "path '" . path2str($path) . "' refer to '$data'\n";
    }

    # path '{a}{b}[0]' refer to 'B0'
    # path '{a}{b}[2]' refer to 'B2'
    # path '{a}{c}[0]' refer to 'C0'

# EXPORT

Nothing is exported by default.

# PATH SYNTAX

Examples:

    '{a}{b}'              # points to b's value
    '{a}{}'               # all values from a's subhash; same for arrays (using empty square brackets)
    '{a}{b,c}'            # b's and c's values
    '{a}{"space inside"}' # key must be quoted unless it is a simple word (single quotes supported as well)
    '{a}{"multi\nline"}'  # same for special characters (if double quoted)
    '{a}{/pattern/mods}'  # regexp keys match (fully supported, except code expressions)
    '{a}{b}[0,1,2,5]'     # 0, 1, 2 and 5 array's items
    '{a}{b}[0..2,5]'      # same, but using ranges
    '{a}{b}[9..0]'        # descending ranges allowed (perl doesn't)
    '{a}{b}(back){c}'     # step back (to previous level)

# SUBROUTINES

## str2path

Convert perl-style string to [Struct::Path](https://metacpan.org/pod/Struct::Path) path structure

    $struct = str2path($string);

## path2str

Convert [Struct::Path](https://metacpan.org/pod/Struct::Path) path structure to perl-style string

    $string = path2str($struct);

# AUTHOR

Michael Samoglyadov, `<mixas at cpan.org>`

# BUGS

Please report any bugs or feature requests to
`bug-struct-path-perlstyle at rt.cpan.org`, or through the web interface at
[http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Struct-Path-PerlStyle](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Struct-Path-PerlStyle). I
will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Struct::Path::PerlStyle

You can also look for information at:

- RT: CPAN's request tracker (report bugs here)

    [http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Path-PerlStyle](http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Path-PerlStyle)

- AnnoCPAN: Annotated CPAN documentation

    [http://annocpan.org/dist/Struct-Path-PerlStyle](http://annocpan.org/dist/Struct-Path-PerlStyle)

- CPAN Ratings

    [http://cpanratings.perl.org/d/Struct-Path-PerlStyle](http://cpanratings.perl.org/d/Struct-Path-PerlStyle)

- Search CPAN

    [http://search.cpan.org/dist/Struct-Path-PerlStyle/](http://search.cpan.org/dist/Struct-Path-PerlStyle/)

# SEE ALSO

[Struct::Path](https://metacpan.org/pod/Struct::Path), [Struct::Diff](https://metacpan.org/pod/Struct::Diff), [perldsc](https://metacpan.org/pod/perldsc), [perldata](https://metacpan.org/pod/perldata)

# LICENSE AND COPYRIGHT

Copyright 2016,2017 Michael Samoglyadov.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See [http://dev.perl.org/licenses/](http://dev.perl.org/licenses/) for more information.


Changes
Revision history for Struct-Path-PerlStyle

0.80    2017-12-22
        - path structs format changed, exported subs renamed to avoid ambigouseness
        - deprecated hook '<<' support removed
        - don't serialize two indexes as range
        - redundant code removed, SYNOPSIS rewritten

0.73    2017-11-05
        - tests for scientific notation or a floating-point numbers fixed

0.72    2017-11-05
        - drop support for perl <= 5.008 (regexp_pattern absent in re pragma)
        - spaces as hash keys delimiters are deprecated (ambigous)
        - hash keys regexps serialization supported
        - hash keys regex parsing fixed and improved
        - parsing/serialization escape sequences in hash keys fixed
        - sane errors for undefs as array indexes on path serialization

0.71    2017-10-15
        - ranges composer in ps_serialize() optimized a bit
        - $_ mangling in ps_parse() fixed

0.70    2017-07-19
        - aliases initial support added (highly experimental)
        - another attempt to fix tests for win32
        - filters/operators renamed to hooks
        - errors log's step index fixed, unification

0.64    2017-06-19
        - '<<' hook renamed to 'back', ('<<' is deprecated now)
        - tests fixed for win32 (I hope =)

0.63    2017-03-31
        - don't quote numbers used as hash keys in ps_serialize()
        - refactoring, small tests improvements
        - tests fixed for dev perl (5.25.9)

0.62    2017-03-01
        - regexp match filter (=~) added
        -croak if unclosed brackets used in path

0.61    2017-02-21
        - steps count corruption fixed in '<<' if closure used more than once
        - 'eq' filter fixed (doesn't croaks now) for undefs

0.60    2017-01-22
        - no backwards compatibility for filters (operators): should be in
            parentheis now (filters still is highly experimental future)
        - filters now may have arguments
        - 'eq', 'defined' and 'not' filters added

0.44    2017-01-12
        - double quotes used in ps_serialize()
        - escape control characters in ps_serialize() (\t,\n,\b and so on)
        - pod corrected

0.43    2017-01-03
        - quote non ASCII keys in ps_serialize()
        - escape/unescape quoting characters
        - tests improved

0.42    2016-12-18
        - Quote key in ps_serialize() unless key is simple word (match /^\w+$/)

0.41    2016-12-06
        - '<' (step back) operator is deprecated ('<<' must be used instead)
        - tests improved

0.40    2016-11-19
        - regexp match for hash keys specification supported

0.33    2016-10-20
        - docs corrected

0.32    2016-10-16
        - bare numbers as hash keys now works correctly
        - get rid of list context for ps_serialize()'s out (ambigous)
        - tests refactored

0.31    2016-09-13
        - ps_serialize() now returns list of serialized steps in list context

0.30    2016-09-12
        - Operators support added to ps_parse()

0.24    2016-09-09
        - passed path corruption fixed (for multiword keys)

0.23    2016-07-17
        - perl 5.10 and below compatibility fixed for ps_serialize()
        - more strict hash definition check for ps_serialize()
        - tests improved

0.22    2016-07-15
        - Keys with spaces/tabs now quoted in ps_serialize()
        - Croak if floating-point array index definition passed to ps_parse()

0.21    2016-07-07
        - Refactoring
        - Tests coverage improved

0.20    2016-06-28
        - Path structure changed according Struct::Path's v0.20

0.05    2016-06-03
        - ranges supported in ps_serialize()

0.04    2016-05-31
        - ps_serialize() implemented

0.03    2016-05-30
        - unstarted/unfinished ranges correctly handled in ps_parse()

0.02    2016-05-27
        - multiple bugfixes, cosmetic changes

0.01    2016-05-19
        - ps_parse() implemented
        - working draft



Makefile.PL
use 5.010;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Struct::Path::PerlStyle',
    AUTHOR           => 'Michael Samoglyadov <mixas@cpan.org>',
    VERSION_FROM     => 'lib/Struct/Path/PerlStyle.pm',
    ABSTRACT_FROM    => 'lib/Struct/Path/PerlStyle.pm',
    LICENSE          => 'Perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.010,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Clone'        => 0,
        'Test::More'   => 0,
    },
    META_MERGE => {
        resources => {
            repository  => 'https://github.com/mr-mixas/Struct-Path-PerlStyle.pm',
        },
    },
    PREREQ_PM => {
        'Carp'              => 0,
        'Safe'              => 0,
        'Text::Balanced'    => 0,
        'Text::ParseWords'  => 0,
        'parent'            => 0,
        're'                => 0.08, # it seems eralier versions has no is_regexp() and regexp_pattern()
        'strict'            => 0,
        'warnings'          => 0
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Struct-Path-PerlStyle-* cover_db' },
);


lib/Struct/Path/PerlStyle/Functions.pm
package Struct::Path::PerlStyle::Functions;

use 5.006;
use strict;
use warnings FATAL => 'all';
use parent 'Exporter';

our @EXPORT_OK = qw(
    back
);

=head1 NAME

Struct::Path::PerlStyle::Functions - Collection handy functions for
L<Struct::Path::PerlStyle> hooks.

=cut

=head1 EXPORT

Nothing is exported by default.

=head1 Functions

=head2 back

Step back count times

    back(3); # go back 3 steps

=cut

sub back {
    my $amount = defined $_[0] ? $_[0] : 1;

    if ($amount > @{$_{path}}) {
        $@ = "Unable to step back such amount of steps";
        return undef;
    }

    splice @{$_{path}}, -$amount;
    splice @{$_{refs}}, -$amount;

    return 1;
}

=head1 AUTHOR

Michael Samoglyadov, C<< <mixas at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-struct-path-native at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Struct-Path-PerlStyle>. I
will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Struct::Path::PerlStyle::Functions

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Path-PerlStyle>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Struct-Path-PerlStyle>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Struct-Path-PerlStyle>

=item * Search CPAN

L<http://search.cpan.org/dist/Struct-Path-PerlStyle/>

=back

=head1 SEE ALSO

L<Struct::Path>, L<Struct::Diff>, L<perldsc>, L<perldata>

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Michael Samoglyadov.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Struct::Path::PerlStyle::Functions


lib/Struct/Path/PerlStyle.pm
package Struct::Path::PerlStyle;

use 5.010;
use strict;
use warnings FATAL => 'all';
use parent 'Exporter';
use utf8;

use Carp 'croak';
use Safe;
use Text::Balanced qw(extract_bracketed extract_quotelike);
use re qw(is_regexp regexp_pattern);

require Struct::Path::PerlStyle::Functions;

our @EXPORT_OK = qw(
    path2str
    str2path
);

=encoding utf8

=head1 NAME

Struct::Path::PerlStyle - Perl-style syntax frontend for L<Struct::Path|Struct::Path>.

=begin html

<a href="https://travis-ci.org/mr-mixas/Struct-Path-PerlStyle.pm"><img src="https://travis-ci.org/mr-mixas/Struct-Path-PerlStyle.pm.svg?branch=master" alt="Travis CI"></a>
<a href='https://coveralls.io/github/mr-mixas/Struct-Path-PerlStyle.pm?branch=master'><img src='https://coveralls.io/repos/github/mr-mixas/Struct-Path-PerlStyle.pm/badge.svg?branch=master' alt='Coverage Status'/></a>
<a href="https://badge.fury.io/pl/Struct-Path-PerlStyle"><img src="https://badge.fury.io/pl/Struct-Path-PerlStyle.svg" alt="CPAN version"></a>

=end html

=head1 VERSION

Version 0.80

=cut

our $VERSION = '0.80';

=head1 SYNOPSIS

    use Struct::Path qw(spath);
    use Struct::Path::PerlStyle qw(path2str str2path);

    my $nested = {
        a => {
            b => ["B0", "B1", "B2"],
            c => ["C0", "C1"],
            d => {},
        },
    };

    my @found = path($nested, str2path('{a}{}[0,2]'), deref => 1, paths => 1);

    while (@found) {
        my $path = shift @found;
        my $data = shift @found;

        print "path '" . path2str($path) . "' refer to '$data'\n";
    }

    # path '{a}{b}[0]' refer to 'B0'
    # path '{a}{b}[2]' refer to 'B2'
    # path '{a}{c}[0]' refer to 'C0'

=head1 EXPORT

Nothing is exported by default.

=head1 PATH SYNTAX

Path is a sequence of 'steps', each represents nested level in the structure.

=head2 Hashes

Like in perl hash keys should be specified using curly brackets

    {}                  # all values from a's subhash
    {foo}               # value for 'foo' key
    {foo,bar}           # slicing: 'foo' and 'bar' values
    {"space inside"}    # key must be quoted unless it is a simple word
    {"multi\nline"}     # special characters interpolated when double quoted
    {/pattern/mods}     # keys regexp match

=head2 Arrays

Square brackets used for array indexes specification

    []                  # all array items
    [9]                 # 9-th element
    [0,1,2,5]           # slicing: 0, 1, 2 and 5 array items
    [0..2,5]            # same, but using ranges
    [9..0]              # descending ranges allowed

=head2 Hooks

Expressions enclosed in parenthesis treated as hooks and evaluated using
L<Safe> compartment. Almost all perl operators and core functions available,
see L<Safe> for more info. Some additional path related functions provided by
L<Struct::Path::PerlStyle::Functions>.

    [](/pattern/mods)           # match array values by regular expression
    []{foo}(eq "bar" && back)   # select hashes which have pair 'foo' => 'bar'

There are two global variables available whithin safe compartment: C<$_> which
refers to value and C<%_> provides current path via key C<path> (in
L<Struct::Path> notation) and structure levels refs stack via key C<refs>.

=head2 Aliases

String in angle brackets is an alias - shortcut mapped into specified sequence
of steps. Aliases resolved recursively, so alias may also map into path with
another aliases.

Aliases map may be defined via global variable

    $Struct::Path::PerlStyle::ALIASES = {
        foo => '{some}{complex}{path}',
        bar => '{and}{one}{more}'
    };

and then

    <foo><bar>      # expands to '{some}{complex}{path}{and}{one}{more}'

or as option for C<str2path>:

    str2path('<foo>', {aliases => {foo => '{long}{path}'}});

=head1 SUBROUTINES

=cut

our $ALIASES;

my %ESCP = (
#    '\\' => '\\\\', # single => double
    '"'  => '\"',
    "\a" => '\a',
    "\b" => '\b',
    "\t" => '\t',
    "\n" => '\n',
    "\f" => '\f',
    "\r" => '\r',
    "\e" => '\e',
);
my $ESCP = join('', sort keys %ESCP);

my %INTP = map { $ESCP{$_} => $_ } keys %ESCP; # swap keys <-> values
my $INTP = join('|', map { "\Q$_\E" } sort keys %INTP);

# $_ will be substituted (if omitted) as first arg if placed on start of
# hook expression
my $COMPL_OPS = join('|', map { "\Q$_\E" }
    qw(< > <= => lt gt le ge == != eq ne ~~ =~));

my $HASH_KEY_CHARS = qr/[\p{Alnum}_\.\-\+]/;

our $HOOK_STRICT = 1;

my $SAFE = Safe->new;
$SAFE->share_from(
    'Struct::Path::PerlStyle::Functions',
    \@Struct::Path::PerlStyle::Functions::EXPORT_OK
);
$SAFE->deny('warn');

my $QR_MAP = {
    ''   => sub { qr/$_[0]/ },
    i    => sub { qr/$_[0]/i },
    m    => sub { qr/$_[0]/m },
    s    => sub { qr/$_[0]/s },
    x    => sub { qr/$_[0]/x },
    im   => sub { qr/$_[0]/im },
    is   => sub { qr/$_[0]/is },
    ix   => sub { qr/$_[0]/ix },
    ms   => sub { qr/$_[0]/ms },
    mx   => sub { qr/$_[0]/mx },
    sx   => sub { qr/$_[0]/sx },
    ims  => sub { qr/$_[0]/ims },
    imx  => sub { qr/$_[0]/imx },
    isx  => sub { qr/$_[0]/isx },
    msx  => sub { qr/$_[0]/msx },
    imsx => sub { qr/$_[0]/imsx },
};

=head2 str2path

Convert perl-style string to L<Struct::Path|Struct::Path> path structure

    $struct = str2path($string);

=cut

sub _push_hash {
    my ($steps, $text) = @_;
    my ($body, $delim, $mods, %step, $token, $type);

    while ($text) {
        ($token, $text, $type, $delim, $body, $mods) =
            (extract_quotelike($text))[0,1,3,4,5,10];

        if (not defined $delim) { # bareword
            push @{$step{K}}, $token = $1
                if ($text =~ s/^\s*($HASH_KEY_CHARS+)//);
        } elsif (!$type and $delim eq '"') {
            $body =~ s/($INTP)/$INTP{$1}/gs; # interpolate
            push @{$step{K}}, $body;
        } elsif (!$type and $delim eq "'") {
            push @{$step{K}}, $body;
        } elsif ($delim eq '/' and !$type or $type eq 'm') {
            $mods = join('', sort(split('', $mods)));
            eval { push @{$step{R}}, $QR_MAP->{$mods}->($body) };
            if ($@) {
                (my $err = $@) =~ s/ at .+//s;
                croak "Step #" . scalar @{$steps} . " $err";
            }
        } else { # things like qr, qw and so on
            substr($text, 0, 0, $token);
            undef $token;
        }

        croak "Unsupported key '$text', step #" . @{$steps}
            if (!defined $token);

        $text =~ s/^\s+//; # discard trailing spaces

        if ($text ne '') {
            if ($text =~ s/^,//) {
                croak "Trailing delimiter at step #" . @{$steps}
                    if ($text eq '');
            } else {
                croak "Delimiter expected before '$text', step #" . @{$steps};
            }
        }
    }

    push @{$steps}, \%step;
}

sub _push_hook {
    my ($steps, $text) = @_;

    # substitute default value if omitted
    $text =~ s/^\s*/\$_ /
        if ($text =~ /^\s*(!\s*|not\s+)*($COMPL_OPS)/);

    my $hook = 'sub {' .
        '$^W = 0; ' .
        'local %_ = ("path", $_[0], "refs", $_[1]); ' .
        'local $_ = (ref $_[1] eq "ARRAY" and @{$_[1]}) ? ${$_[1]->[-1]} : undef; ' .
        $text .
    '}';

    open (local *STDERR,'>', \(my $stderr)); # catch compilation errors

    unless ($hook = $SAFE->reval($hook, $HOOK_STRICT)) {
        if ($stderr) {
            $stderr =~ s/ at \(eval \d+\) .+//s;
            $stderr = " ($stderr)";
        } else {
            $stderr = "";
        }

        (my $err = $@) =~ s/ at \(eval \d+\) .+//s;
        croak "Failed to eval hook '$text': $err, step #" . @{$steps} . $stderr;
    }

    push @{$steps}, $hook;
}

sub _push_list {
    my ($steps, $text) = @_;
    my (@range, @step);

    for my $i (split /\s*,\s*/, $text, -1) {
        @range = grep {
            croak "Incorrect array index '$i', step #" . @{$steps}
                unless (eval { $_ == int($_) });
        } ($i =~ /^\s*(-?\d+)\s*\.\.\s*(-?\d+)\s*$/) ? ($1, $2) : $i;

        push @step, $range[0] < $range[-1]
            ? $range[0] .. $range[-1]
            : reverse $range[-1] .. $range[0];
    }

    push @{$steps}, \@step;
}

sub str2path($;$) {
    my ($path, $opts) = @_;

    croak "Undefined path passed" unless (defined $path);

    local $ALIASES = $opts->{aliases} if (exists $opts->{aliases});

    my (@steps, $step, $type);

    while ($path) {
        # separated match: to be able to have another brackets inside;
        # currently mostly for hooks, for example: '( $x > $y )'
        for ('{"}', '["]', '(")', '<">') {
            ($step, $path) = extract_bracketed($path, $_, '');
            last if ($step);
        }

        croak "Unsupported thing in the path, step #" . @steps . ": '$path'"
            unless ($step);

        $type = substr $step,  0, 1, ''; # remove leading bracket
                substr $step, -1, 1, ''; # remove trailing bracket

        if ($type eq '{') {
            _push_hash(\@steps, $step);
        } elsif ($type eq '[') {
            _push_list(\@steps, $step);
        } elsif ($type eq '(') {
            _push_hook(\@steps, $step);
        } else { # <>
            if (exists $ALIASES->{$step}) {
                substr $path, 0, 0, $ALIASES->{$step};
                redo;
            }

            croak "Unknown alias '$step'";
        }
    }

    return \@steps;
}

=head2 path2str

Convert L<Struct::Path|Struct::Path> path structure to perl-style string

    $string = path2str($struct);

=cut

sub path2str($) {
    my $path = shift;

    croak "Arrayref expected for path" unless (ref $path eq 'ARRAY');
    my $out = '';
    my $sc = 0; # step counter

    for my $step (@{$path}) {
        my @items;

        if (ref $step eq 'ARRAY') {
            for my $i (@{$step}) {
                croak "Incorrect array index '" . ($i // 'undef') . "', step #$sc"
                    unless (eval { int($i) == $i });
                if (@items and (
                    $items[-1][0] < $i and $items[-1][-1] == $i - 1 or   # ascending
                    $items[-1][0] > $i and $items[-1][-1] == $i + 1      # descending
                )) {
                    $items[-1][1] = $i; # update range
                } else {
                    push @items, [$i]; # new range
                }
            }

            for (@{items}) {
                $_ = abs($_->[0] - $_->[-1]) < 2
                    ? join(',', @{$_})
                    : "$_->[0]..$_->[-1]"
            }

            $out .= "[" . join(",", @{items}) . "]";
        } elsif (ref $step eq 'HASH') {
            my $types = [ grep { exists $step->{$_} } qw(K R) ];
            if (keys %{$step} != @{$types}) {
                $types = { map { $_, 1 } @{$types} };
                my @errs = grep { !exists $types->{$_} } sort keys %{$step};
                croak "Unsupported hash definition (" .
                    join(',', @errs) . "), step #$sc"
            }

            if (exists $step->{K}) {
                croak "Unsupported hash keys definition, step #$sc"
                    unless (ref $step->{K} eq 'ARRAY');

                for my $k (@{$step->{K}}) {
                    croak "Unsupported hash key type 'undef', step #$sc"
                        unless (defined $k);
                    croak "Unsupported hash key type '@{[ref $k]}', step #$sc"
                        if (ref $k);

                    push @items, $k;

                    unless ($k =~ /^$HASH_KEY_CHARS+$/) {
                        $items[-1] =~ s/([\Q$ESCP\E])/$ESCP{$1}/gs;    # escape
                        $items[-1] = qq("$items[-1]");                 # quote
                    }
                }
            }

            if (exists $step->{R}) {
                croak "Unsupported hash regexps definition, step #$sc"
                    unless (ref $step->{R} eq 'ARRAY');

                for my $r (@{$step->{R}}) {
                    croak "Regexp expected for regexps item, step #$sc"
                        unless (is_regexp($r));

                    my ($patt, $mods) = regexp_pattern($r);
                    $mods =~ s/[dlu]//g; # for Perl's internal use (c) perlre
                    push @items, "/$patt/$mods";
                }
            }

            $out .= "{" . join(",", @items) . "}";
        } else {
            croak "Unsupported thing in the path, step #$sc";
        }
        $sc++;
    }

    return $out;
}

=head1 AUTHOR

Michael Samoglyadov, C<< <mixas at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-struct-path-perlstyle at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Struct-Path-PerlStyle>. I
will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Struct::Path::PerlStyle

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Struct-Path-PerlStyle>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Struct-Path-PerlStyle>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Struct-Path-PerlStyle>

=item * Search CPAN

L<http://search.cpan.org/dist/Struct-Path-PerlStyle/>

=back

=head1 SEE ALSO

L<Struct::Path>, L<Struct::Path::JsonPointer>, L<Struct::Diff>
L<perldsc>, L<perldata>, L<Safe>

=head1 LICENSE AND COPYRIGHT

Copyright 2016-2018 Michael Samoglyadov.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1; # End of Struct::Path::PerlStyle
